### PR TITLE
Adding the TdsWrite Stream tests

### PR DESCRIPTION
**Pull request by saurabh500**
Adding the TdsWrite Stream tests

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs [hunk 1]
```diff
@@ -0,0 +1,292 @@
+﻿using System;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:1]: Add file header

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+using System.ComponentModel;
+using System.IO;
+using System.Threading;
+using Microsoft.Data.SqlClientX.IO;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.UnitTests.IO
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:8]: Can fix namespace to 'UnitTests.IO' ?

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+                Assert.Contains(header.Status, _packetStatuses);
+                if (header.Status == TdsEnums.ST_BATCH)
+                {
+                    Assert.Equal(packetSize, header.Length);
+                    // When Batch packets are sent out, they are expected to be completely full to the packet size.
+                    Assert.Equal(lengthRead, packetSize);
+                }
+
+                // Match content 
+                int contentLength = lengthRead - 8;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:88]: Assert contentLength before slice?

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+                // Assert the packet type is either ST_EOM or ST_BATCH.
+                Assert.Contains(header.Status, _packetStatuses);
+                if (header.Status == TdsEnums.ST_BATCH)
+                {
+                    Assert.Equal(packetSize, header.Length);
+                    // When Batch packets are sent out, they are expected to be completely full to the packet size.
+                    Assert.Equal(lengthRead, packetSize);
+                }
+                // Match content 
+                int contentLength = lengthRead - 8;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:138]: Same here.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+                byte[] expectedData = inputBuffer.AsSpan().Slice(matchOffset, contentLength).ToArray();
+                Assert.Equal(expectedData, header.Content);
+                // Move the match offset to the next packet.
+                matchOffset += contentLength;
+            } while(header?.Status != TdsEnums.ST_EOM);
+        }
+
+        [Fact]
+        public async void WriteStream_WriteSinglePacketAsync()
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:147]: Nit: Please add inline comments where possible, so intent of test is readable..

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs (hunk 1 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:1]: Sigh... Thanks for pointint it out.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs (hunk 2 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:8]: Ya

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs (hunk 3 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:88]: Ya

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs (hunk 5 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:147]: Will do. Thanks.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs (hunk 2 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:8]: @cheenamalhotra  looking at the tests in FunctionalTests and Manual tests, we are following the fully qualified namespace where we are prefixing the NS with Microsoft.Data.SqlClient. I would like to stick to the pattern.

File: build.proj [hunk 6]
```diff
@@ -61,8 +61,8 @@
     <FunctionalTests Include="**/FunctionalTests/Microsoft.Data.SqlClient.Tests.csproj" />
     <FunctionalTestsProj Include="**/FunctionalTests/Microsoft.Data.SqlClient.Tests.csproj" />
 
-    <UnitTests Include="**/UnitTests/Microsoft.Data.SqlClient.UnitTests.csproj" />
-    <UnitTestsProj Include="**/UnitTests/Microsoft.Data.SqlClient.UnitTests.csproj" />
+    <UnitTests Include="**/UnitTests/Microsoft.Data.SqlClient.NetCore.UnitTests.csproj" />
```
Comment (saurabh500) [inline build.proj:64]: This should have been done in the last PR of UT merge. But tackling it here.

**Comment by saurabh500** [comment]
@cheenamalhotra  addressed all the PR comments.

**Review by cheenamalhotra — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+using System.Threading;
+using Microsoft.Data.SqlClientX.IO;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.UnitTests.IO
+{
+    /// <summary>
+    /// Unit tests for the TdsWriteStream class.
+    /// </summary>
+    public class TdsWriteStreamTest
```
Comment (benrr101) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:16]: So I'm definitely glad there's a renewed commitment to unit tests, and I'm glad to see that they are already paying off by finding bugs. I'd push a bit on going a little further :)

Organizing tests into Arrange / Act / Assert makes test much more readable.

I tend to go method by method for the tests, and try to go through each scenario that might happen. It does result in duplication in a bunch of cases, but it does help with verifying edge cases are covered. This is one thing I haven't gotten ai to generate completely yet. For instance, I would write a few tests `Write_LessThanBuffer` (writes less bytes than the buffer, verifies no flush happens), `Write_EqualToBuffer` (writes up to end of buffer, verifies no flush happens?), `Write_LargetThanBuffer` (writes entire buffer and then some, verifies flush happens). Likewise with Flush, I'd write `Flush_BufferEmpty`, `Flush_BufferPartial`, `Flush_BufferFull`. 

I won't hold this review since this is definitely much better than no tests. But I'd also be happy to add more tests to enhance it.

**Review by benrr101 — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs (hunk 7 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/IO/TdsWriteStreamTest.cs:16]: Yeah, there are lot more things that can be done for UTs
Feel free to build on this, I am going to merge this PR. I am happy that I have been able to take care of critical bug fixes due to UTs.

Then I am going to get my Tds Read stream, since that unblocks Cheena.
